paddle/framework/backward.cc

```diff
@@ -0,0 +1,190 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License. */
+
+#include "paddle/framework/backward.h"
+#include <list>
+#include "paddle/framework/net.h"
+#include "paddle/framework/op_registry.h"
+
+namespace paddle {
+namespace framework {
+
+static bool AllInSet(const std::vector<std::string>& names,
+                     const std::string& suffix,
+                     const std::unordered_set<std::string>& set) {
+  for (auto& name : names) {
+    if (set.find(name + suffix) == set.end()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+static std::vector<size_t> InSetIdx(
+    const std::vector<std::string>& names, const std::string& suffix,
+    const std::unordered_set<std::string>& set) {
+  std::vector<size_t> ret_val;
+  ret_val.reserve(names.size());
+  for (size_t i = 0; i < names.size(); ++i) {
+    if (set.find(names[i] + suffix) != set.end()) {
+      ret_val.push_back(i);
+    }
+  }
+  return ret_val;
+}
+
+static std::shared_ptr<OperatorBase> EmptyOp() {
```

Comment:
It seems that `NOP` is only used in `Backward` and should not be used by users. So just write `NOP` here.